libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,404 @@
+	values[YulString{zero}] = AssignedValue{&zeroLiteral, {}};
+	values[YulString{one}] = AssignedValue{&oneLiteral, {}};
+	values[YulString{thirtyTwo}] = AssignedValue{&thirtyTwoLiteral, {}};
+
+	bool const ignoreMemory = MSizeFinder::containsMSize(_context.dialect, _ast);
+	UnusedStoreEliminator rse{
+		_context.dialect,
+		functionSideEffects,
+		ControlFlowSideEffectsCollector{_context.dialect, _ast}.functionSideEffectsNamed(),
+		values,
+		ignoreMemory
+	};
+	rse(_ast);
+	rse.changeUndecidedTo(State::Unused, Location::Memory);
+	rse.changeUndecidedTo(State::Used, Location::Storage);
+	rse.scheduleUnusedForDeletion();
+
+	StatementRemover remover(rse.m_pendingRemovals);
+	remover(_ast);
+}
+
+void UnusedStoreEliminator::operator()(FunctionCall const& _functionCall)
+{
+	UnusedStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+		changeUndecidedTo(sideEffects.canTerminate ? State::Used : State::Unused, Location::Storage);
+	}
+}
+
+void UnusedStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	UnusedStoreBase::operator()(_functionDefinition);
+}
+
+
+void UnusedStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void UnusedStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	UnusedStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	// We determine if this is a store instruction without additional side-effects
+	// both by querying a combination of semantic information and by listing the instructions.
+	// This way the assert below should be triggered on any change.
+	using evmasm::SemanticInformation;
+	bool isStorageWrite = (*instruction == Instruction::SSTORE);
+	bool isMemoryWrite =
+		*instruction == Instruction::EXTCODECOPY ||
+		*instruction == Instruction::CODECOPY ||
+		*instruction == Instruction::CALLDATACOPY ||
+		*instruction == Instruction::RETURNDATACOPY ||
+		*instruction == Instruction::MSTORE ||
+		*instruction == Instruction::MSTORE8;
+	bool isWriteToBeHandled =
```

Comment:
```suggestion
	bool isCandidateForRemoval =
```
?
That's what this really is, right?

Reply:
It also took me a second to see why this is done on the statement level, but we can of course only remove statements anyways... I wonder if in the future, once we'd summarize stores for yul functions, it'll be worthwhile to consider splitting a function into its write part and the computation of its return value part, though...
I.e. try to transform
```
function f(a) -> r { sstore(0, a) r := add(a,1) }
sstore(1, f(42))
sstore(0, 0)
```
```
function f(a) -> r { sstore(0, a) r := add(a,1) }
function f_pure_part(a) -> r { r := add(a,1) }
sstore(1, f_pure_part(42))
sstore(0, 0)
```
and thus ultimately do this on the expression level instead...
But not sure if that likely happened a lot and right now probably just a distraction to consider...